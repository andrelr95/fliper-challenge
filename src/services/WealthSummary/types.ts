interface WealthSummary {
  id: number;
  model: string;
  year: number;
  stock: number;
}

export interface WealthSummaryData {
  wealthSummary: WealthSummary[];
}
