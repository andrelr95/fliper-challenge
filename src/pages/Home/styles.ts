import styled from 'styled-components';
import Text from 'components/Text';

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
`;

export const InvestedMoneyText = styled(Text).attrs(() => ({
  color: 'darkGray',
  type: 'medium16',
}))`
  margin-bottom: 8px;
`;
