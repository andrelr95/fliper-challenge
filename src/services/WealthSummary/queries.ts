import { gql } from '@apollo/client';

export const GET_WEALTH_SUMMARY = gql`
  {
    wealthSummary {
      id
      cdi
      gain
      hasHistory
      profitability
      total
    }
  }
`;
