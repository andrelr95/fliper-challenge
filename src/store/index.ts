import { configureStore } from '@reduxjs/toolkit';

import wealthSummary from './modules/WealthSummary/wealth-summary.store';

export const store = configureStore({
  reducer: {
    wealthSummary,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
