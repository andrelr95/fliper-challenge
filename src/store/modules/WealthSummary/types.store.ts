export interface WealthSummary {
  id: number;
  total: number;
  profitability: number;
  hasHistory: boolean;
  gain: number;
  cdi: number;
}

export interface WealthSummaryState {
  payload: WealthSummary;
  loading: string;
}
