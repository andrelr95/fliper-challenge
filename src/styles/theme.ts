/* eslint-disable @typescript-eslint/no-explicit-any */
import { DefaultTheme as DefaultStyledTheme } from 'styled-components';

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: { [key in keyof typeof themeColors]: string };
    fontTypes: { [key in keyof typeof fontTypes]: any };
    shadows: { [key in keyof typeof shadowColors]: string };
  }
}

export const themeColors = {
  white: '#FFF',
  grayBlue: '#9AA3BC',
  grayBlue2: '#E8EBF6',
  darkGray: '#606377',
  darkBlue: '#3B5CB8',
};

export const shadowColors = {
  light: '1px 1px 1px rgba(0, 0, 0, 0.16)',
};

export const fontTypes = {
  medium16: {
    fontFamily: 'Milliard Medium',
    fontSize: '16px',
    letterSpacing: '0.48px',
    lineHeight: '20px',
  },
  bold16: {
    fontFamily: 'Metropolis Bold',
    fontSize: '16px',
    letterSpacing: '0.48px',
    lineHeight: '20px',
  },
  xbold18: {
    fontFamily: 'Metropolis Extra Bold',
    fontSize: '18px',
    letterSpacing: '0.54px',
    lineHeight: '22px',
  },
  xbold22: {
    fontFamily: 'Metropolis Extra Bold',
    fontSize: '22px',
    letterSpacing: '0.66px',
    lineHeight: '30px',
  },
  xbold25: {
    fontFamily: 'Metropolis Extra Bold',
    fontSize: '25px',
    letterSpacing: '0.75px',
    lineHeight: '30px',
  },
};

const theme: DefaultStyledTheme = {
  colors: themeColors,
  fontTypes,
  shadows: shadowColors,
};

export default theme;
