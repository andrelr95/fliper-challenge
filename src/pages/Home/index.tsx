/* eslint-disable no-console */
import React, { useEffect } from 'react';
import { useHistory } from 'react-router';
import { useQuery } from '@apollo/client';
import 'styled-components/macro';

import Card from 'components/Card';
import Button from 'components/Button';

import { GET_WEALTH_SUMMARY } from 'services/WealthSummary/queries';
import { WealthSummaryData } from 'services/WealthSummary/types';

import Text from 'components/Text';
import Icon from 'components/Icon';
import {
  formatNumberToCurrencyString,
  formatNumberToPercentage,
} from 'helpers';

import * as S from './styles';

const Home = () => {
  const history = useHistory();
  const { loading, data, error } =
    useQuery<WealthSummaryData>(GET_WEALTH_SUMMARY);

  useEffect(() => {
    if (error) {
      console.log('card error', { error });
    }
  }, [error]);

  return (
    <S.Wrapper>
      <Card
        loading={loading}
        header={
          <>
            <Text color="darkBlue" type="xbold25">
              Seu resumo
            </Text>
            <Icon name="options" onClick={() => console.log('OPTIONS')} />
          </>
        }
        footer={
          <Button
            color="primary"
            type="button"
            onClick={() => history.push('/details')}
          >
            <Text type="bold16" color="darkBlue">
              ver mais
            </Text>
          </Button>
        }
      >
        {data && (
          <S.Content>
            <S.InvestedMoneyWrapper>
              <S.InvestedMoneyLabel>Valor investido</S.InvestedMoneyLabel>
              <S.InvestedMoneyValue>
                {formatNumberToCurrencyString(data.wealthSummary[0].total)}
              </S.InvestedMoneyValue>
            </S.InvestedMoneyWrapper>
            <S.DetailsWrapper>
              <S.Detail>
                <S.DetailLabel>Rentabilidade/mês</S.DetailLabel>
                <S.DetailValue>
                  {formatNumberToPercentage(
                    data.wealthSummary[0].profitability,
                    3,
                  )}
                </S.DetailValue>
              </S.Detail>
              <S.Detail>
                <S.DetailLabel>CDI</S.DetailLabel>
                <S.DetailValue>
                  {formatNumberToPercentage(data.wealthSummary[0].cdi, 2)}
                </S.DetailValue>
              </S.Detail>
              <S.Detail>
                <S.DetailLabel>Ganho/mês</S.DetailLabel>
                <S.DetailValue>
                  {formatNumberToCurrencyString(data.wealthSummary[0].gain)}
                </S.DetailValue>
              </S.Detail>
            </S.DetailsWrapper>
          </S.Content>
        )}
      </Card>
    </S.Wrapper>
  );
};

export default Home;
