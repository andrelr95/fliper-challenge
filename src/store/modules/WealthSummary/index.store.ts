/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { WealthSummaryState } from './types.store';

const INITIAL_STATE: WealthSummaryState = {
  payload: {
    cdi: 0,
    gain: 0,
    hasHistory: false,
    id: 0,
    profitability: 0,
    total: 0,
  },
  loading: 'idle',
};

export const wealthSummarySlice = createSlice({
  name: 'counter',
  initialState: INITIAL_STATE,
  reducers: {
    storeData: (state, action: PayloadAction<WealthSummaryState>) => {
      state = { ...action.payload };
    },
  },
});

export const { storeData } = wealthSummarySlice.actions;
export default wealthSummarySlice.reducer;
