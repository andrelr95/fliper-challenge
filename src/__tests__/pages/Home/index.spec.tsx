import React from 'react';
import renderer from 'react-test-renderer';
import createStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { MockedProvider } from '@apollo/client/testing';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import theme from 'styles/theme';
import { WealthSummaryState } from 'store/modules/WealthSummary/wealth-summary.store';

import Home from 'pages/Home';
import { GET_WEALTH_SUMMARY } from 'services/WealthSummary/queries';

const INITIAL_STATE: WealthSummaryState = {
  cdi: 0,
  gain: 0,
  hasHistory: false,
  id: 0,
  profitability: 0,
  total: 0,
};

// const mocks = [
//   {
//     request: {
//       query: GET_WEALTH_SUMMARY,
//     },
//     result: {
//       data: {
//         wealthSummary: [
//           {
//             id: 2,
//             cdi: 3.45678,
//             gain: 1833.23,
//             hasHistory: true,
//             profitability: 2.76789,
//             total: 3200876,
//             __typename: 'wealthSummary',
//           },
//           {
//             id: 1,
//             cdi: 2,
//             gain: 400,
//             hasHistory: false,
//             profitability: 4.5,
//             total: 6470.1,
//             __typename: 'wealthSummary',
//           },
//         ],
//       },
//     },
//   },
// ]; // We'll fill this in next

const mockStore = createStore([]);
const store = mockStore(INITIAL_STATE);

describe('Home', () => {
  it('Renders correctly', async () => {
    const { act } = renderer;
    const wealthSummaryMock = {
      request: {
        query: GET_WEALTH_SUMMARY,
        variables: {},
      },
      result: {
        data: {
          wealthSummary: [
            {
              id: 2,
              cdi: 3.45678,
              gain: 1833.23,
              hasHistory: true,
              profitability: 2.76789,
              total: 3200876,
              __typename: 'wealthSummary',
            },
            {
              id: 1,
              cdi: 2,
              gain: 400,
              hasHistory: false,
              profitability: 4.5,
              total: 6470.1,
              __typename: 'wealthSummary',
            },
          ],
        },
      },
    };
    const tree = renderer
      .create(
        <Provider store={store}>
          <MockedProvider mocks={[wealthSummaryMock]} addTypename={false}>
            <ThemeProvider theme={theme}>
              <MemoryRouter>
                <Home />
              </MemoryRouter>
            </ThemeProvider>
          </MockedProvider>
        </Provider>,
      )
      .toJSON();
    await act(() => {
      return new Promise(resolve => setTimeout(resolve, 0));
    });
    expect(tree).toMatchSnapshot();
  });
});
