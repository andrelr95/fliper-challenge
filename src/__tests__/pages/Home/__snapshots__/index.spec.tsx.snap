// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Renders correctly 1`] = `
<div
  className="sc-iCoGMd ewpiJz"
>
  <div
    className="sc-bdnxRM kigoMP"
    id="card"
  >
    <div
      data-testid="card-loading-skeleton"
    >
      <div
        className="css-1521o73-SkeletonTheme"
      >
        <span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton-sc-bxkw17-0 BAJYD css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 30,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
        </span>
        <span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton2-sc-bxkw17-1 fwWcuw css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 90,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
        </span>
        <span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton3-sc-bxkw17-2 clrdvF css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 24,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton3-sc-bxkw17-2 clrdvF css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 24,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton3-sc-bxkw17-2 clrdvF css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 24,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
        </span>
        <span>
          <span
            className="react-loading-skeleton Loading___StyledSkeleton4-sc-bxkw17-3 dGKXdM css-1eg0sfo-skeletonStyles-Skeleton"
            style={
              Object {
                "height": 60,
                "width": "100%",
              }
            }
          >
            ‌
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
